### ISSUE TITLE
Future Direction - Multiplatform, no JVM dependencies

### ISSUE THREAD

**Issue by thomasnield** [thomasnield/kotlin-statistics#29]
As Kotlin-Statistics continues to grow in functionality, less of it is depending on Apache Commons Math. 

It would be great if we got rid of Apache Commons Math as a dependency altogether, and help Kotlin-Statistics go multi-platform and behave more as a standard library. 

This will primarily require descriptive statistics and distributions to be implemented from scratch. It's possible some linear algebra library/implementation will be needed to back such functionality. There will also be an emphasis on discrete and continuous probability functionality, and integrating it intuitively with the Kotlin language and stdlib. 

Timeline-wise, I think it would be best to time with a major release of Kotlin and [have some much-needed features implemented](https://github.com/thomasnield/KEEP/blob/master/proposals/stdlib/exclusive-ranges.md). 

Any other thoughts or ideas until then can be discussed here.

**Comment by kurtseebauer** [comment]
why implement the needed stuff from scratch? commons-math is apache licenced, so you could copy the code, (auto-)convert it to kotlin and maybe get rid of some unwanted depencencies (logging?), if there are any. 
The best case of course would be to convert the whole commons-math library to kotlin/multiplatform and publish it to maven, but I have no idea how much work that would mean.

**Comment by thomasnield** [comment]
@kurtseebauer I'm investigating such approaches. Ports are fine, but like you said there are a lot of dependencies and there is manual refactoring that has to be done. I have been looking at ojAlgo for inspiration as well, as it has zero dependencies and is pretty solid. http://www.ojalgo.org/

**Comment by thomasnield** [comment]
Moving this to a scope document: 
https://github.com/thomasnield/kotlin-statistics/blob/master/version_2_scope.md

**Comment by Groostav** [comment]
Any update here @thomasnield ?

I'm looking to have some very smart non-systems people do some fairly serious math in native code. So far they've been working in fortran. I would like this madness to end. Kotlin native + kotlin statistics represents a very compelling platform.

**Comment by thomasnield** [comment]
@Groostav yes and no. There has been some discussions with @altavir to start using kmath to be the backbone to Kotlin Statistics 2.0. 

Obviously been busy these days, with full-time job, a startup initiative, O'Reilly, and getting married. But play around with this first: 

https://github.com/mipt-npm/kmath

**Comment by altavir** [comment]
@thomasnield My congratulations!

@Groostav Thank you for your interest. A little status report. We've 80% done with migrating statistics core to the multiplatform. Additionally I've implemented a streaming API which allows to compute statistical quantities not only for fixed arrays, but also for data which is being loaded or keeps coming. I got slowed down by porting Commons Math percentile algorithm and then was overwhelmed by urgent work. @pklimai recently joined the work and written a temporary solution for percentiles (do anybody use them anyway?) so I hope that the work will continue soon.

We currently do not support native platforms due to more complicated build and deploy process (it should work though). I am not sure that Kotlin native is a good solution if you are not dead locked on some native code interop. It is not quite fast at the moment and has much less deploy and debug options. Could you please elaborate on your task parameters?

**Comment by thomasnield** [comment]
@altavier thank you :) Also regarding percentiles, I use percentiles all the time at work specifically for on-time performance measurements and planning. 

@everyone After I get settled down with married life I am hoping to have more time to focus on this. I don't have any conferences and events distracting me from the remainder of the year. The only thing that could dominate my time is if my startup gets funding and I actually have to start doing work.

**Comment by kastork** [comment]
"After I get settled down with married life" ...

Good luck with that.  :)

But on a more serious note on the subject of RNGs and distributions.  https://dst.lbl.gov/ACSSoftware/colt/ has a lot to offer and is quite rigorous with respect to the maths.

**Comment by altavir** [comment]
Colt is quite old and is not maintained since 2004. You should not use it. On the other hand Apache Commons Rng is released recently and contains some well designed tools for random number generation. I do not have a lot of time right now, but I've added basic support for distributions in Dev branch of kmath and some bindings for older Commons math  implementations: https://github.com/mipt-npm/kmath/blob/dev/kmath-commons/src/main/kotlin/scientifik/kmath/commons/prob/CommonsDistribution.kt. if you need it, I can release it.also I can make smaller module, which uses Rng, not full maths.

**Comment by joreilly** [comment]
Has there been any more thoughts about making this work in Kotlin Multiplatform shared code?

**Comment by altavir** [comment]
It actually does: https://github.com/thomasnield/kotlin-statistics/tree/mpp-dev/src. I am not sure if @thomasnield has time to publish it right now, but you can always fork it.

**Comment by thomasnield** [comment]
@altavir @joreilly hey folks, yes sorry I've been rather busy with some other initiatives right now and bandwidth is really thin. Feel free to fork and support I wish I had more time to keep this going ;(
